fs-storage/src/jni_file_storage.rs

```diff
@@ -0,0 +1,185 @@
+    // This is the interface to the JVM that we'll call the majority of our
+    // methods on.
+    use jni::JNIEnv;
+
+    // These objects are what you should use as arguments to your native
+    // function. They carry extra lifetime information to prevent them escaping
+    // this context and getting used after being GC'd.
+    use jni::objects::{JClass, JString, JValue};
+
+    // This is just a pointer. We'll be returning it from our function. We
+    // can't return one of the objects with lifetime information because the
+    // lifetime checker won't let us.
+    use jni::sys::{jboolean, jlong, jobject};
+
+    use crate::base_storage::BaseStorage;
+
+    use crate::file_storage::FileStorage;
+
+    impl FileStorage<String, String> {
+        fn from_jlong<'a>(value: jlong) -> &'a mut Self {
+            unsafe { &mut *(value as *mut FileStorage<String, String>) }
+        }
+    }
+
+    #[no_mangle]
+    pub extern "system" fn Java_FileStorage_create<'local>(
+        mut env: JNIEnv<'local>,
+        _class: JClass,
+        label: JString<'local>,
+        path: JString<'local>,
+    ) -> jlong {
+        let label: String = env
+            .get_string(&label)
+            .expect("Couldn't get label!")
+            .into();
+        let path: String = env
+            .get_string(&path)
+            .expect("Couldn't get path!")
+            .into();
+
+        let file_storage: FileStorage<String, String> =
+            FileStorage::new(label, Path::new(&path));
+        Box::into_raw(Box::new(file_storage)) as jlong
+    }
+
+    #[no_mangle]
+    pub extern "system" fn Java_FileStorage_set<'local>(
+        mut env: JNIEnv<'local>,
+        _class: JClass,
+        id: JString<'local>,
+        value: JString<'local>,
+        file_storage_ptr: jlong,
+    ) {
+        let id: String = env.get_string(&id).expect("msg").into();
+        let value: String = env.get_string(&value).expect("msg").into();
+
+        FileStorage::from_jlong(file_storage_ptr).set(id, value);
+    }
+
+    #[no_mangle]
+    pub extern "system" fn Java_FileStorage_remove<'local>(
+        mut env: JNIEnv<'local>,
+        _class: JClass,
+        id: JString<'local>,
+        file_storage_ptr: jlong,
+    ) {
+        let id: String = env.get_string(&id).unwrap().into();
+        FileStorage::from_jlong(file_storage_ptr)
+            .remove(&id)
+            .unwrap_or_else(|err| {
+                env.fatal_error(err.to_string());
+            });
+    }
+
+    #[no_mangle]
+    pub extern "system" fn Java_FileStorage_needsSyncing(
+        env: JNIEnv<'_>,
+        _class: JClass,
+        file_storage_ptr: jlong,
+    ) -> jboolean {
+        match FileStorage::from_jlong(file_storage_ptr).needs_syncing() {
+            Ok(updated) => updated as jboolean,
+            Err(err) => {
+                env.fatal_error(err.to_string());
+            }
+        }
+    }
+
+    #[no_mangle]
+    pub extern "system" fn Java_FileStorage_readFS(
```

Comment:
Resorting to Java's `LinkedHashMap` could be fine for this prototype, but I think we should try the approach with writing our own bindings to `BTreeMap`.